### PR TITLE
Add shaping functions

### PR DESCRIPTION
**Pull request by henridf**
This commit adds the shaping 'primitive' functions cast(), fill(),
crop(), and order() along with fit() (composition of fill and crop)
and shape (composition of all four primitive functions). 

A description of what these functions do is in #1783; the ztests of this commit also provide some illustration.

I will expand the test set after #1932 is done, since a lot of shaping assignments will use that.

### CODE REVIEW

File: expr/shaper.go [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zson"
+)
+
+// A ShaperTransform represents one of the different transforms that a
+// shaper can apply.
+type ShaperTransform int
+
+const (
+	Cast = ShaperTransform(iota)
```
Comment (nwt) [inline expr/shaper.go:17]: NIt: Use conventional form.
```suggestion
	Cast ShaperTransform = iota
```

File: expr/shaper.go [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+	Fill
+	Crop
+	Order
+)
+
+type step int
+
+const (
+	copyPrimitive = step(iota) // copy field fromIndex from input record
```
Comment (nwt) [inline expr/shaper.go:26]: NIt: Use conventional form.
```suggestion
	copyPrimitive step = iota // copy field fromIndex from input record
```

File: expr/shaper.go [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+	copyContainer
+	castPrimitive // cast field fromIndex from input record
+	null          // write null
+	record        // record into record below us
+)
+
+// A shapeBuilder is a recursive data structure encoding a series of
+// copy/cast steps to be carried out over an input record.
+type shapeBuilder struct {
+	op        step
```
Comment (nwt) [inline expr/shaper.go:36]: I think you should change the type name to `op`.

File: compiler/expr.go [hunk 4]
```diff
@@ -258,6 +258,52 @@ func compileCutter(zctx *resolver.Context, node ast.FunctionCall) (*expr.Cutter,
...
+		return map[expr.ShaperTransform]struct{}{expr.Crop: struct{}{}}
+	case "order":
+		return map[expr.ShaperTransform]struct{}{expr.Order: struct{}{}}
+	case "shape":
+		return map[expr.ShaperTransform]struct{}{expr.Crop: struct{}{}, expr.Fill: struct{}{}, expr.Order: struct{}{}, expr.Cast: struct{}{}}
+	default:
+		return nil
+	}
+
+}
```
Comment (nwt) [inline compiler/expr.go:279]: Maps feel heavy here. Maybe use flags?
```go
func shaperOps(name string) ShaperTransform {
	switch name {
	case "cast":
		return expr.Cast
	case "fill":
		return expr.Fill
	case "fit":
		return expr.Crop | expr.Fill
	case "crop":
		return expr.Crop
	case "order":
		return expr.Order
	case "shape":
		return expr.Crop | expr.Fill | expr.Orde | expr.Cast
	default:
		return 0
	}
}
```

File: compiler/expr.go [hunk 5]
```diff
@@ -258,6 +258,52 @@ func compileCutter(zctx *resolver.Context, node ast.FunctionCall) (*expr.Cutter,
...
+
+func isShaperFunc(name string) bool {
+	switch name {
+	case "cast", "fill", "fit", "crop", "order", "shape":
+		return true
+	}
+	return false
```
Comment (nwt) [inline compiler/expr.go:286]:
```suggestion
	return shaperOps(name) != nil
```

File: zng/type.go [hunk 6]
```diff
@@ -337,6 +337,17 @@ func IsUnionType(typ Type) bool {
...
 
+func IsRecordType(typ Type) bool {
+	switch typ := typ.(type) {
+	case *TypeAlias:
+		return IsRecordType(typ.Type)
+	case *TypeRecord:
+		return true
+	default:
+		return false
+	}
```
Comment (nwt) [inline zng/type.go:348]:
```suggestion
	_, ok := AliasedType(typ).(*TypeRecord)
	return ok
```

File: compiler/expr.go (hunk 4 above)
Comment (henridf) [reply compiler/expr.go:279]: Yes!

File: zng/type.go (hunk 6 above)
Comment (henridf) [reply zng/type.go:348]: 👍

File: compiler/expr.go (hunk 5 above)
Comment (henridf) [reply compiler/expr.go:286]: 👍

File: expr/shaper.go (hunk 3 above)
Comment (henridf) [reply expr/shaper.go:36]: done in 4cd9d91

**Review by nwt — APPROVED** [review]

**Pull request merged**